WordPress/Classes/ViewRelated/Reader/ReaderTagCardCellViewModel.swift

```diff
@@ -0,0 +1,100 @@
+            guard let post = try? ContextManager.shared.mainContext.existingObject(with: objectID) as? ReaderPost,
+                  let cell = collectionView.dequeueReusableCell(withReuseIdentifier: ReaderTagCell.classNameWithoutNamespaces(), for: indexPath) as? ReaderTagCell else {
+                return UICollectionViewCell()
+            }
+            cell.configure(parent: self?.parentViewController,
+                           post: post,
+                           isLoggedIn: self?.isLoggedIn ?? AccountHelper.isLoggedIn)
+            return cell
+        }
+    }()
+
+    private lazy var resultsController: NSFetchedResultsController<ReaderPost> = {
+        let fetchRequest = NSFetchRequest<ReaderPost>(entityName: ReaderPost.classNameWithoutNamespaces())
+        fetchRequest.sortDescriptors = [NSSortDescriptor(key: "sortRank", ascending: false)]
+        fetchRequest.fetchLimit = Constants.displayPostLimit
+        let resultsController = NSFetchedResultsController<ReaderPost>(fetchRequest: fetchRequest,
+                                                           managedObjectContext: ContextManager.shared.mainContext,
+                                                           sectionNameKeyPath: nil,
+                                                           cacheName: nil)
+        resultsController.delegate = self
+        return resultsController
+    }()
+
+    init(parent: UIViewController?, tag: ReaderTagTopic, collectionView: UICollectionView?, isLoggedIn: Bool, cellSize: @escaping @autoclosure () -> CGSize?) {
+        self.parentViewController = parent
+        self.slug = tag.slug
+        self.collectionView = collectionView
+        self.isLoggedIn = isLoggedIn
+        self.cellSize = cellSize
+
+        super.init()
+
+        resultsController.fetchRequest.predicate = NSPredicate(format: "topic = %@ AND isSiteBlocked = NO", tag)
+        collectionView?.delegate = self
+    }
+
+    func fetchTagTopics() {
+        try? resultsController.performFetch()
+    }
+
+    func onTagButtonTapped() {
+        let controller = ReaderStreamViewController.controllerWithTagSlug(slug)
+        parentViewController?.navigationController?.pushViewController(controller, animated: true)
+    }
+
+    struct Constants {
+        static let displayPostLimit = 10
+    }
+
+}
+
+// MARK: - NSFetchedResultsControllerDelegate
+
+extension ReaderTagCardCellViewModel: NSFetchedResultsControllerDelegate {
+
+    func controller(_ controller: NSFetchedResultsController<any NSFetchRequestResult>,
+                    didChangeContentWith snapshot: NSDiffableDataSourceSnapshotReference) {
+        dataSource?.apply(snapshot as Snapshot, animatingDifferences: false)
+    }
+
+}
+
+// MARK: - UICollectionViewDelegate
+
+extension ReaderTagCardCellViewModel: UICollectionViewDelegate {
+
+    func collectionView(_ collectionView: UICollectionView, didSelectItemAt indexPath: IndexPath) {
+        let post = resultsController.object(at: indexPath)
```

Comment:
**Optional suggestion** — This may seem a bit paranoid 😅 , but we could probably add a boundary check here to prevent `object(at:)` from raising an exception:

```suggestion
        guard let sectionInfo = resultsController.sections[safe: indexPath.section],
              indexPath.row < sectionInfo.numberOfObjects else {
            return
        }
        let post = resultsController.object(at: indexPath)
```

I'll leave it up to you to decide.

Reply:
I'd be quite surprised if we hit an out of bounds here considering it's a `didSelectItemAt` call, but I've also seen weirder things happen 😆. I've added it in https://github.com/wordpress-mobile/WordPress-iOS/pull/23057/commits/532544575925f2058006d5893aa7efb9efc29420, thanks!